### PR TITLE
Fix test_comprehensive__softmax_backward_data_cpu_float16

### PR DESCRIPTION
**Pull request by Flamefire**
The test fails with `Exception: expected scalar type Float but found Half` as the inputs are converted to Float but the `input_dtype` argument is `float16`.

There is a comment referencing #63057 at https://github.com/pytorch/pytorch/blob/41d7be3f48df402837ba1bd1e436a99df448ae7d/aten/src/ATen/native/SoftMax.cpp#L92-L105

I'm not fully sure at which point things break: The inputs (grad and output) are converted to Float, input_dtype is passed as-is, but `softmax_backward_cpu_out` ignores `input_dtype`

This might be a regression introduced by 7255847ca2f272aa761ea8692c2b4826e53ef47b which (re)enabled reference_in_float for CPU possibly with some conflict with a newly added test or code change that used to rely on that.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by Flamefire** [comment]
Already done in the meantime in 53e958ddf88a
